### PR TITLE
 Store the final status of fields in `componentt`

### PR DESCRIPTION
**Pull request by thomasspriggs**
Store the final status of fields in `componentt`

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by kroening — CHANGES_REQUESTED** [review]
The test needs to be added to the Makefile.

File: src/util/std_types.h [hunk 1]
```diff
@@ -223,6 +223,16 @@ class struct_union_typet:public typet
     {
       return set(ID_C_is_padding, is_padding);
     }
+
+    bool get_is_final() const
+    {
+      return get_bool(ID_final);
+    }
+
+    void set_is_final(const bool is_final)
```
Comment (LAJW) [inline src/util/std_types.h:198]: The name makes me cry.

**Review by LAJW — APPROVED** [review]

File: src/util/std_types.h (hunk 1 above)
Comment (tautschnig) [reply src/util/std_types.h:198]: So ... what would you recommend instead?

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 2]
```diff
@@ -753,6 +753,8 @@ void java_bytecode_convert_classt::convert(
     else
       component.set_access(ID_default);
 
+    component.set_is_final(f.is_final);
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:756]: :bulb: If the class is opaque the `is_final` flag should be set to true.

**Comment by tautschnig** [comment]
@thomasspriggs Asking for debugging reasons: It seems the PR comment is missing the bits of information that .github/pull_request_template.md should have added. Were those not presented to you when creating the PR, or did you just create this PR quite a while ago? It seemed to work in #2975.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp (hunk 2 above)
Comment (allredj) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:756]: After discussion with @thomasspriggs, we worked out that dealing with fields of opaque classes may not be as straightforward than I thought. So my comment can be ignored.

**Comment by thomasspriggs** [comment]
@tautschnig Yes, I just raised this PR, there was a template added and I deleted it.

**Comment by thomasspriggs** [comment]
@kroening I have added the unit test to the makefile, please re-review.

**Comment by tautschnig** [comment]
> [...] there was a template added and I deleted it.

Which part of the template suggested that it can safely be deleted?

**Comment by thomasspriggs** [comment]
@tautschnig It superficially appeared to be standard operating procedures for working on a PR, rather than something which needed to be filled out and left in a PR description. I assumed that if it couldn't be safely deleted someone would complain. :)

File: src/util/std_types.h (hunk 1 above)
Comment (peterschrammel) [reply src/util/std_types.h:198]: How about `set_final`?

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_field.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      THEN("The the final status of the classes fields should be correct.")
+      {
+        const symbolt &class_symbol =
+          symbol_table.lookup_ref("java::ClassWithFields");
+        const java_class_typet &java_class =
+          to_java_class_type(class_symbol.type);
+        REQUIRE(java_class.get_component("final1").get_is_final());
+        REQUIRE(java_class.get_component("final2").get_is_final());
+        REQUIRE(java_class.get_component("final3").get_is_final());
+        REQUIRE(!java_class.get_component("nonFinal1").get_is_final());
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_field.cpp:35]: :pick: Prefer `REQUIRE_FALSE`

File: jbmc/unit/java_bytecode/java_bytecode_parser/ClassWithFields.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+
+class ClassWithFields {
+  final boolean final1 = true;
+  final Boolean final2 = false;
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/ClassWithFields.java:4]: :bulb: I'd have probably picked different accessibility and/or different ways of setting them (i.e. in the constructor). You might also like to add tests for fields that hide inherited fields

**Review by thk123 — APPROVED** [review]
@tautschnig The addition of the template was not communicated to me so sorry about that 

@thomasspriggs suggest opening a new PR to copy the template in and add to this PR.

File: src/util/std_types.h (hunk 1 above)
Comment (allredj) [reply src/util/std_types.h:198]: FYI we already use `(s|g)et_is_final` for methods. So whatever we choose we should remain consistent.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 5874f53e2475ba0d4941d2322bfb18164f8d3fa5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85166768

**Comment by tautschnig** [comment]
> @tautschnig The addition of the template was not communicated to me so sorry about that
>
> @thomasspriggs suggest opening a new PR to copy the template in and add to this PR.

I have taken the liberty to just copy&paste from .github/pull_request_template.md (the template got added via PR #2973).

**Review by kroening — APPROVED** [review]

**Pull request merged**